toolkit/tools/scheduler/schedulerutils/buildworker.go

```diff
@@ -99,14 +107,22 @@ func BuildNodeWorker(channels *BuildChannels, agent buildagents.BuildAgent, grap
 	// it when we pick up the next request
 	for req, cancelled := selectNextBuildRequest(channels); !cancelled && req != nil; req, cancelled = selectNextBuildRequest(channels) {
 		res := &BuildResult{
-			Node:           req.Node,
-			AncillaryNodes: req.AncillaryNodes,
-			WasDelta:       req.IsDelta,
+			Node:            req.Node,
+			AncillaryNodes:  req.AncillaryNodes,
+			WasDelta:        req.IsDelta,
+			ActualFreshness: req.ExpectedFreshness,
 		}
 
 		switch req.Node.Type {
 		case pkggraph.TypeLocalBuild:
-			res.UsedCache, res.Skipped, res.BuiltFiles, res.LogFile, res.Err = buildBuildNode(req.Node, req.PkgGraph, graphMutex, agent, req.CanUseCache, buildAttempts, checkAttempts, ignoredPackages)
+			var hadMissingFiles bool
+			res.UsedCache, res.Skipped, hadMissingFiles, res.BuiltFiles, res.LogFile, res.Err = buildBuildNode(req.Node, req.PkgGraph, graphMutex, agent, req.CanUseCache, buildAttempts, checkAttempts, ignoredPackages)
+			if hadMissingFiles {
+				res.ActualFreshness = NodeFreshnessRebuildRequired
+				logger.Log.Debugf("Resetting freshness to %d for %s", res.ActualFreshness, res.Node.FriendlyName())
```

Comment:
Since `NodeFreshnessRebuildRequired` is a special-case constant, I think we might get more info about explicitly saying why we're resetting the freshness instead of printing the value.

```suggestion
				logger.Log.Debugf("Resetting freshness for %s due to missing files.", res.ActualFreshness, res.Node.FriendlyName())
```